python/pylibraft/pylibraft/cluster/kmeans.pyx

```diff
@@ -220,3 +229,165 @@ def compute_new_centroids(X,
+        if batch_samples is not None:
+            self.c_obj.batch_samples = batch_samples
+        if batch_centroids is not None:
+            self.c_obj.batch_centroids = batch_centroids
+        if inertia_check is not None:
+            self.c_obj.inertia_check = inertia_check
+
+        # TODO: distance metric/ verbosity level (?) / rng state
+
+    @property
+    def n_clusters(self):
+        return self.c_obj.n_clusters
+
+    @property
+    def max_iter(self):
+        return self.c_obj.max_iter
+
+    @property
+    def tol(self):
+        return self.c_obj.tol
+
+    @property
+    def init(self):
+        return InitMethod(self.c_obj.init)
+
+    @property
+    def oversampling_factor(self):
+        return self.c_obj.oversampling_factor
+
+    @property
+    def batch_samples(self):
+        return self.c_obj.batch_samples
+
+    @property
+    def batch_centroids(self):
+        return self.c_obj.batch_centroids
+
+    @property
+    def inertia_check(self):
+        return self.c_obj.inertia_check
+
+KMeansOutput = namedtuple("FitOutput", "centroids inertia n_iter")
+
+
+@auto_sync_handle
+def fit(
+    KMeansParams params, X, centroids=None, sample_weights=None, handle=None
+):
+
+    """
+    Fit kmeans
+
+    Parameters
+    ----------
+
+    X : Input CUDA array interface compliant matrix shape (m, k)
+    centroids : Optional writable CUDA array interface compliant matrix
+                shape (n_clusters, k)
+    sample_weights : Optional input CUDA array interface compliant matrix shape
+                     (n_clusters, 1) default: None
+    {handle_docstring}
+
+    Examples
+    --------
+
+    .. code-block:: python
+
+        import cupy as cp
+
+        from pylibraft.cluster.kmeans import fit, KMeansParams
+
+        n_samples = 5000
+        n_features = 50
+        n_clusters = 3
+
+        X = cp.random.random_sample((n_samples, n_features),
+                                      dtype=cp.float32)
+
+        params = KMeansParams(n_clusters=n_clusters)
+        centroids, inertia, n_iter = fit(params, X)
+    """
+    cdef handle_t *h = <handle_t*><size_t>handle.getHandle()
+
+    x_cai = X.__cuda_array_interface__
+    dtype = np.dtype(x_cai["typestr"])
+
+    cdef float f_inertia = 0.0
+    cdef double d_inertia = 0.0
+    cdef int n_iter = 0
+
+    cdef optional[device_vector_view[const double, int]] d_sample_weights
+    cdef optional[device_vector_view[const float, int]] f_sample_weights
+
+    if centroids is None:
+        centroids_shape = (params.n_clusters, x_cai["shape"][1])
+        centroids = device_ndarray.empty(centroids_shape, dtype=dtype)
+
+    if dtype == np.float64:
+        if sample_weights is not None:
+            d_sample_weights = const_device_vector_view_from_array[double](
+                sample_weights, <double*>NULL)
+
+        cpp_kmeans.fit(
+            deref(h),
+            params.c_obj,
+            const_device_matrix_view_from_array[double](X, <double*>NULL),
```

Comment:
I've changed in the last commit - I don't think its too bad myself

We could also maybe add some code to the cai_wrapper to make this a little cleaner (like convert to pyx, then cdef a function to return the data as a `uintptr_t` to avoid the extra casts needed here etc). What do you think?

Reply:
Yeah that sounds good to me. Maybe have cai_wrapper.data return the uintptr_t.

Reply:
We could also probably add a `size` property to `cai_wrapper` that would automatically return the product of the shape.